squeal-postgresql/src/Squeal/PostgreSQL/Binary.hs

```diff
@@ -513,6 +513,105 @@ instance FromValue pg y
+  fromArray = Decoding.nullableValueArray (fromValue @('PGchar 1))
+instance FromArray ('NotNull 'PGtext) Strict.Text where
+  fromArray = Decoding.valueArray (fromValue @'PGtext)
+instance FromArray ('Null 'PGtext) (Maybe Strict.Text) where
+  fromArray = Decoding.nullableValueArray (fromValue @'PGtext)
+instance FromArray ('NotNull 'PGtext) Lazy.Text where
+  fromArray = Decoding.valueArray (fromValue @'PGtext)
+instance FromArray ('Null 'PGtext) (Maybe Lazy.Text) where
+  fromArray = Decoding.nullableValueArray (fromValue @'PGtext)
+instance FromArray ('NotNull 'PGbytea) Strict.ByteString where
+  fromArray = Decoding.valueArray (fromValue @'PGbytea)
+instance FromArray ('Null 'PGbytea) (Maybe Strict.ByteString) where
+  fromArray = Decoding.nullableValueArray (fromValue @'PGbytea)
+instance FromArray ('NotNull 'PGbytea) Lazy.ByteString where
+  fromArray = Decoding.valueArray (fromValue @'PGbytea)
+instance FromArray ('Null 'PGbytea) (Maybe Lazy.ByteString) where
+  fromArray = Decoding.nullableValueArray (fromValue @'PGbytea)
+instance FromArray ('NotNull 'PGdate) Day where
+  fromArray = Decoding.valueArray (fromValue @'PGdate)
+instance FromArray ('Null 'PGdate) (Maybe Day) where
+  fromArray = Decoding.nullableValueArray (fromValue @'PGdate)
+instance FromArray ('NotNull 'PGtime) TimeOfDay where
+  fromArray = Decoding.valueArray (fromValue @'PGtime)
+instance FromArray ('Null 'PGtime) (Maybe TimeOfDay) where
+  fromArray = Decoding.nullableValueArray (fromValue @'PGtime)
+instance FromArray ('NotNull 'PGtimetz) (TimeOfDay, TimeZone) where
+  fromArray = Decoding.valueArray (fromValue @'PGtimetz)
+instance FromArray ('Null 'PGtimetz) (Maybe (TimeOfDay, TimeZone)) where
+  fromArray = Decoding.nullableValueArray (fromValue @'PGtimetz)
+instance FromArray ('NotNull 'PGtimestamp) LocalTime where
+  fromArray = Decoding.valueArray (fromValue @'PGtimestamp)
+instance FromArray ('Null 'PGtimestamp) (Maybe LocalTime) where
+  fromArray = Decoding.nullableValueArray (fromValue @'PGtimestamp)
+instance FromArray ('NotNull 'PGtimestamptz) UTCTime where
+  fromArray = Decoding.valueArray (fromValue @'PGtimestamptz)
+instance FromArray ('Null 'PGtimestamptz) (Maybe UTCTime) where
+  fromArray = Decoding.nullableValueArray (fromValue @'PGtimestamptz)
+instance FromArray ('NotNull 'PGinterval) DiffTime where
+  fromArray = Decoding.valueArray (fromValue @'PGinterval)
+instance FromArray ('Null 'PGinterval) (Maybe DiffTime) where
+  fromArray = Decoding.nullableValueArray (fromValue @'PGinterval)
+instance FromArray ('NotNull 'PGjson) Value where
+  fromArray = Decoding.valueArray (fromValue @'PGjson)
+instance FromArray ('Null 'PGjson) (Maybe Value) where
+  fromArray = Decoding.nullableValueArray (fromValue @'PGjson)
+instance FromArray ('NotNull 'PGjsonb) Value where
+  fromArray = Decoding.valueArray (fromValue @'PGjsonb)
+instance FromArray ('Null 'PGjsonb) (Maybe Value) where
+  fromArray = Decoding.nullableValueArray (fromValue @'PGjsonb)
```

Comment:
Yeah, it's necessary for correct binary instances for nested arrays, no way to avoid it as far as I can tell. I played around with type families but in the end it felt heavy weight and didn't buy anything really. I cleaned up some of the boilerplate using overlapping instances for now. It's still not perfect but it covers the cases a user will likely want.